File: leetcode2/0050. Pow(x, n).py
class Solution:
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        rep=0
        if x>=0 and n>=0:
            rep=self.submyPow(x,n)
        elif x>=0 and n<0:
            rep=1/self.submyPow(x,-n)
        elif x<0 and n%2==0 and n>=0:
            rep=self.submyPow(-x,n)
        elif x<0 and n%2==0 and n<0:
            rep=1/self.submyPow(-x,n)
        elif x<0 and n%2!=0 and n>0:
            rep=-self.submyPow(-x,n)
        else:
            rep=-1/self.submyPow(-x,n)
        return rep
    
    def submyPow(self,x,n):
        rep=1
        while n>0:
            cur=x
            cnt=1
            while cnt*2<n:
                cur=cur*cur
                cnt=cnt+cnt
            n=n-cnt
            rep*=cur
        return rep

File: archives/C#/0050. Pow(x, n).cs
public class Solution {
    public double MyPow(double x, int n) {
        double m=n;
        if(n<0){
            m=-m;
            x=1/x;
        }
        double result=1;
        while(m>0){
            double tmp=x;
            int cnt=1;
            while(cnt<m-cnt){
                tmp*=tmp;
                cnt+=cnt;
            }
            m=m-cnt;
            result*=tmp;
        }
        return result;
    }
}
